### PR TITLE
[effects] Add way to register effectul op

### PR DESCRIPTION
**Pull request by angelayi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122348
* #122347

This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.

### CODE REVIEW

File: test/higher_order_ops/test_with_effects.py [hunk 1]
```diff
@@ -199,6 +213,130 @@ def f(x):
 
         res.sum().backward()
 
+    def test_register_effectful_custom_op(self):
+        lib = torch.library.Library("mylib", "FRAGMENT")  # noqa: TOR901
+        try:
```
Comment (zou3519) [inline test/higher_order_ops/test_with_effects.py:218]: use torch.library._scoped_library

File: test/higher_order_ops/test_with_effects.py [hunk 2]
```diff
@@ -199,6 +213,130 @@ def f(x):
...
+                "(Tensor x, str prefix) -> ()",
+                lib=lib,
+            )
+
+            # global variable to store the recorded tensor and prefix.
+            recorded_dict = {}
+
+            # Pytorch custorm op implementation
+            @torch.library.impl(
+                "mylib::record_scalar_tensor", "CompositeExplicitAutograd"
```
Comment (zou3519) [inline test/higher_order_ops/test_with_effects.py:233]: need to pass lib=lib

File: test/higher_order_ops/test_with_effects.py [hunk 3]
```diff
@@ -199,6 +213,130 @@ def f(x):
...
+            )
+            def record_scalar_tensor(x, prefix):
+                recorded_dict[prefix] = x.clone()
+                return
+
+            # Meta function of the custom op
+            @torch.library.impl_abstract("mylib::record_scalar_tensor")
```
Comment (zou3519) [inline test/higher_order_ops/test_with_effects.py:240]: lib=lib

**Review by zou3519 — APPROVED** [review]

File: test/higher_order_ops/test_with_effects.py (hunk 1 above)
Comment (angelayi) [reply test/higher_order_ops/test_with_effects.py:218]: I just copied it from [other test cases](https://www.internalfb.com/code/search?q=regex%3Aon%20repo%3Afbcode%20torch.library.Library.*%23%20noqa%3A%20TOR901) 😅 but now I know!

**Pull request closed without merging**